File: 0386-lexicographical-numbers/0386-lexicographical-numbers.ts
function lexicalOrder(n: number): number[] {
  const ans: number[] = new Array(n).fill(0);
  let d = 1;
  for (let i = 0; i < n; i++) {
    ans[i] = d;
    do {
      if (d * 10 <= n) {
        d *= 10;
      } else {
        while (d % 10 === 9) d = Math.floor(d / 10)
        d += 1;
      }
    } while (d > n);
  } 
  return ans;
};

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.go
func lexicalOrder(n int) []int {
  ans := make([]int, n)
  ans[0] = 1
  for i, d := 1, 1; i < n; i++ {
    if d * 10 <= n {
      d *= 10
    } else {
      for d % 10 == 9 || d >= n {
        d /= 10
      }
      d++
    }
    ans[i] = d
  } 
  return ans
}